classes/Village.php

```diff
@@ -129,4 +135,287 @@ public function getRelations(System $system, string $village_id): array {
+            $resource_history[$resource_id]['collected'] = (int)$result['collected'];
+            // get claimed
+            $result = $system->db->query("SELECT COUNT(*) as 'claimed' FROM `loot` WHERE `resource_id` = {$resource_id} AND `claimed_village_id` = {$village_id} AND `claimed_time` > {$time}");
+            $result = $system->db->fetch($result);
+            $resource_history[$resource_id]['claimed'] = (int)$result['claimed'];
+            // get lost (lol)
+            // can determine resources lost based on difference of produced and collected
+            $resource_history[$resource_id]['lost'] = $resource_history[$resource_id]['produced'] - $resource_history[$resource_id]['collected'];
+            // factor in resources which haven't been collected from regions
+            $result = $system->db->query("SELECT SUM(`resource_count`) as 'quantity' FROM `region_locations` INNER JOIN `regions` on `regions`.`region_id` = `region_locations`.`region_id` WHERE `resource_id` = {$resource_id} AND `village` = {$village_id}");
+            $result = $system->db->fetch($result);
+            $resource_history[$resource_id]['lost'] -= (int) $result['quantity'];
+            // get upkeep - WIP
+            $resource_history[$resource_id]['spent'] = 0;
+        }
+        // extra step to calculate lost resources: factor in resources which haven't been collected from caravans
+        $caravan_result = $system->db->query("SELECT * FROM `caravans` where `village_id` = {$village_id}");
+        $caravan_result = $system->db->fetch_all($caravan_result);
+        foreach ($caravan_result as $caravan) {
+            $caravan_resources = json_decode($caravan['resources'], true);
+            // reduce resource from lost count
+            foreach ($caravan_resources as $resource_id => $quantity) {
+                $resource_history[$resource_id]['lost'] -= $quantity;
+            }
+        }
+
+        return $resource_history;
+    }
+
+    /**
+     * @return Village
+     */
+    public static function getVillageByID(System $system, int $village_id): Village
+    {
+        $result = $system->db->query(
+            "SELECT * FROM `villages`
+                WHERE `villages`.`village_id`='{$village_id}'"
+        );
+        $result = $system->db->fetch($result);
+        return new Village($system, village_row: $result);
+    }
+
+    /**
+     * @return array
+     */
+    public static function getClans(System $system, string $village): array
+    {
+        $result = $system->db->query("SELECT * FROM `clans` WHERE `village` = '{$village}'");
+        $result = $system->db->fetch_all($result);
+        return $result;
+    }
+
+    /**
+     * @return array
+     */
+    public static function getPlayerSeat(System $system, int $player_id): array
+    {
+        $result = $system->db->query("SELECT * FROM `village_seats` WHERE `user_id` = {$player_id} AND `seat_end` IS NULL LIMIT 1");
+        $result = $system->db->fetch($result);
```

Comment:
Maybe we could use the seat info DTO here too?

Reply:
Complete agree, will do.